src/tbb/market.cpp

```diff
@@ -364,22 +363,28 @@ arena* market::arena_in_need ( arena_list_type* arenas, arena* hint ) {
             } while ( arenas[curr_priority_level].empty() );
             it = arenas[curr_priority_level].begin();
         }
-        if( a.num_workers_active() < a.my_num_workers_allotted.load(std::memory_order_relaxed) ) {
+
+        if ( &a == current &&
+             a.my_pool_state.load(std::memory_order_relaxed) != arena::SNAPSHOT_EMPTY &&
+             a.num_workers_active() - 1 < a.my_num_workers_allotted.load(std::memory_order_relaxed) )
+        {
+            return &a;
+        } else if ( a.num_workers_active() < a.my_num_workers_allotted.load(std::memory_order_relaxed) ) {
             a.my_references += arena::ref_worker;
             return &a;
         }
```

Comment:
Do not we have logical race here? E.g. 
the first check:
`&a == current` - `true`
`a.my_pool_state.load(std::memory_order_relaxed) != arena::SNAPSHOT_EMPTY` - `true`
`a.num_workers_active() - 1 < a.my_num_workers_allotted.load(std::memory_order_relaxed)` - `false`
the second check (some threads left):
`a.num_workers_active() < a.my_num_workers_allotted.load(std::memory_order_relaxed)` - `true`
As a result, we add the second reference to the same arena.